cpp/src/arrow/util/rle_encoding_test.cc

```diff
@@ -553,47 +853,178 @@ TEST(BitRle, Overflow) {
+            static_cast<int32_t>(data_remaining->null_count()),
+            data_remaining->null_bitmap_data(), data_remaining->offset());
+      } else {
+        auto* out = values_read.data() + requested_read_count;
+        read = decoder.GetBatchSpaced(
+            to_read, static_cast<int32_t>(data_remaining->null_count()),
+            data_remaining->null_bitmap_data(), data_remaining->offset(), out);
+      }
+    } else {
+      if (dict) {
+        auto* out = dict_read.data() + requested_read_count;
+        read = decoder.GetBatchWithDict(
+            dict->raw_values(), static_cast<int32_t>(dict->length()), out, to_read);
+      } else {
+        auto* out = values_read.data() + requested_read_count;
+        read = decoder.GetBatch(out, to_read);
+      }
+    }
+    ASSERT_EQ(read, to_read) << "Decoder did not read as many values as requested";
 
-  for (int64_t i = 0; i < num_values; ++i) {
-    if (data.IsValid(i)) {
-      if (values_read[i] != values[i]) {
-        FAIL() << "Index " << i << " read " << values_read[i] << " but should be "
-               << values[i];
+    actual_read_count += read;
+    requested_read_count += to_read;
+  }
+  EXPECT_EQ(requested_read_count, data_size) << "This test logic is wrong";
+  EXPECT_EQ(actual_read_count, data_size) << "Total number of values read is off";
+
+  // Verify the round trip: encoded-decoded values must equal the original one
+  for (int64_t i = 0; i < data_size; ++i) {
+    if (data.IsValid(i) || !kSpaced) {
+      if (dict) {
+        EXPECT_EQ(dict_read.at(i), dict->Value(data_values[i]))
+            << "Encoded then decoded and mapped value at position " << i << " ("
+            << values_read[i] << ") differs from original value (" << data_values[i]
+            << " mapped to " << dict->Value(data_values[i]) << ")";
+      } else {
+        EXPECT_EQ(values_read.at(i), data_values[i])
+            << "Encoded then decoded value at position " << i << " (" << values_read.at(i)
+            << ") differs from original value (" << data_values[i] << ")";
       }
     }
   }
 }
 
+template <typename T>
+struct DataTestRleBitPackedRandomPart {
+  using value_type = T;
+
+  value_type max;
+  int32_t size;
+  double null_probability;
+};
+
+template <typename T>
+struct DataTestRleBitPackedRepeatPart {
+  using value_type = T;
+
+  value_type value;
+  int32_t size;
+};
+
+template <typename T>
+struct DataTestRleBitPackedNullPart {
+  using value_type = T;
+
+  int32_t size;
+};
+
+template <typename T>
+struct DataTestRleBitPacked {
+  using value_type = T;
+  using ArrowType = typename arrow::CTypeTraits<value_type>::ArrowType;
+  using RandomPart = DataTestRleBitPackedRandomPart<value_type>;
+  using RepeatPart = DataTestRleBitPackedRepeatPart<value_type>;
+  using NullPart = DataTestRleBitPackedNullPart<value_type>;
+
+  std::vector<std::variant<RandomPart, RepeatPart, NullPart>> parts;
```

Comment:
Can we define a type alias for this `std::variant`? This will make things more readable.